gtdynamics/optimizer/ConstraintManifold.cpp

```diff
@@ -0,0 +1,275 @@
+    Values linearization_point;
+    linearization_point.insert(key, values.at(key));
+    graph.emplace_shared<LinearContainerFactor>(linear_factor,
+                                                linearization_point);
+  }
+  gtsam::LevenbergMarquardtOptimizer optimizer(graph, init_values_cc,
+                                               params_->lm_params);
+  return optimizer.optimize();
+  // auto result = 
+  // if (graph.error(result) > 1e-4) {
+  //   std::cout << "cannot retract exactly tproj\n";
+  // }
+  // return result;
+}
+
+/* ************************************************************************* */
+void ConstraintManifold::compute_basis_kernel() {
+  auto linear_graph = cc_->merit_graph.linearize(values_);
+  JacobianFactor combined(*linear_graph);
+  auto augmented = combined.augmentedJacobian();
+  Matrix A = augmented.leftCols(augmented.cols() - 1);  // m x n
+  // Vector b = augmented.col(augmented.cols() - 1);
+  Eigen::FullPivLU<Eigen::MatrixXd> lu(A);
+  basis_ = lu.kernel();  // n x n-m
+  size_t position = 0;
+  for (const Key& key : combined.keys()) {
+    size_t var_dim = values_.at(key).dim();
+    var_dim_[key] = var_dim;
+    var_location_[key] = position;
+    position += var_dim;
+  }
+}
+
+/* ************************************************************************* */
+void ConstraintManifold::compute_basis_specify_variables() {
+  size_t basis_dim = 0;
+  for (const Key& key : basis_keys_) {
+    basis_dim += values_.at(key).dim();
+  }
+  if (basis_dim != dim()) {
+    std::cerr << "specified basis have wrong dimensions\n";
+  }
+
+  auto linear_graph = cc_->merit_graph.linearize(values_);
+  auto full_ordering =
+      Ordering::ColamdConstrainedLast(*linear_graph, basis_keys_);
+  Ordering ordering = full_ordering;
+  for (size_t i = 0; i < basis_keys_.size(); i++) {
+    ordering.pop_back();
+  }
+  auto elim_result = linear_graph->eliminatePartialSequential(ordering);
+  auto bayes_net = elim_result.first;
+
+  size_t position = 0;
+  for (const Key& key : full_ordering) {
+    size_t var_dim = values_.at(key).dim();
+    var_dim_[key] = var_dim;
+    var_location_[key] = position;
+    position += var_dim;
+  }
+
+  basis_ = Matrix::Zero(position, dim());
+
+  size_t col_idx = 0;
+  for (const Key& basis_key : basis_keys_) {
+    auto dim = values_.at(basis_key).dim();
+    for (size_t dim_idx = 0; dim_idx < dim; dim_idx++) {
+      // construct basis values
+      VectorValues sol_missing;
+      for (const Key& key : basis_keys_) {
+        Vector vec = Vector::Zero(values_.at(key).dim());
+        if (key == basis_key) {
+          vec(dim_idx) = 1;
+        }
+        sol_missing.insert(key, vec);
+      }
+
+      VectorValues result = bayes_net->optimize(sol_missing);
+
+      for (const Key& key : full_ordering) {
+        basis_.block(var_location_.at(key), col_idx, var_dim_.at(key), 1) =
+            result.at(key);
+      }
+      col_idx += 1;
+    }
+  }
+}
+
+
```

Comment:
Please run formatting on all files. There are extra spaces and newlines which the formatter will remove.

Reply:
Done